Source/Core/Core/AchievementManager.cpp

```diff
@@ -246,6 +252,50 @@ void AchievementManager::DoFrame()
   }
 }
 
+bool AchievementManager::CanPause()
+{
+  u32 frames_to_next_pause = 0;
+  bool can_pause = rc_client_can_pause(m_client, &frames_to_next_pause);
+  if (!can_pause)
+  {
+    OSD::AddMessage("Cannot spam pausing in hardcore mode.", OSD::Duration::VERY_LONG,
+                    OSD::Color::RED);
+    OSD::AddMessage(
+        fmt::format("Can pause in {} seconds.", static_cast<float>(frames_to_next_pause) / 60),
+        OSD::Duration::VERY_LONG, OSD::Color::RED);
+  }
+  return can_pause;
+}
+
+void AchievementManager::DoIdle()
+{
+  std::thread([this]() {
+    while (true)
+    {
+      Common::SleepCurrentThread(1000);
+      {
+        std::lock_guard lg{m_lock};
+        if (!m_system || Core::GetState(*m_system) != Core::State::Paused)
+          return;
+        if (!m_background_execution_allowed)
+          return;
+        if (!m_client || !IsGameLoaded())
+          return;
+      }
+      Core::QueueHostJob([this](Core::System& system) {
```

Comment:
Would be nice to have a comment in this function about why we're moving this to the host thread. Weird workarounds like this are the kind of thing that are in most need of comments, as the comments help people in the future understand not only why we have this code to begin with but also whether they can safely remove it.